app/src/main/java/com/realityexpander/tasky/core/util/getRealPathFromURI.kt

```diff
@@ -0,0 +1,152 @@
+                        }
+                        "audio" -> {
+                            contentUri = MediaStore.Audio.Media.EXTERNAL_CONTENT_URI
+                        }
+                    }
+                    val selection = "_id=?"
+                    val selectionArgs = arrayOf(split[1])
+                    return getDataColumn(context, contentUri, selection, selectionArgs)
+                }
+            }
+        }
+        "content".equals(uri.scheme, ignoreCase = true) -> {
+            // Return the remote address
+            return if (isGooglePhotosUri(uri)) uri.lastPathSegment else getDataColumn(context, uri, null, null)
+        }
+        "file".equals(uri.scheme, ignoreCase = true) -> {
+            return uri.path
+        }
+    }
+    return null
+}
+
+fun getDataColumn(context: Context, uri: Uri?, selection: String?,
+                  selectionArgs: Array<String>?): String? {
+    var cursor: Cursor? = null
+    val column = "_data"
+    val projection = arrayOf(
+            column
+    )
+    try {
+        if (uri == null) return null
+        cursor = context.contentResolver.query(uri, projection, selection, selectionArgs,
+                null)
+        if (cursor != null && cursor.moveToFirst()) {
+            val index = cursor.getColumnIndexOrThrow(column)
+            return cursor.getString(index)
+        }
+    } finally {
+        cursor?.close()
+    }
+    return null
+}
+
+
+fun getFilePath(context: Context, uri: Uri?): String? {
+    var cursor: Cursor? = null
+    val projection = arrayOf(
+            MediaStore.MediaColumns.DISPLAY_NAME
+    )
+    try {
+        if (uri == null) return null
+        cursor = context.contentResolver.query(uri, projection, null, null,
+                null)
+        if (cursor != null && cursor.moveToFirst()) {
+            val index = cursor.getColumnIndexOrThrow(MediaStore.MediaColumns.DISPLAY_NAME)
+            return cursor.getString(index)
+        }
+    } finally {
+        cursor?.close()
+    }
+    return null
+}
+
+/**
+ * @param uri The Uri to check.
+ * @return Whether the Uri authority is ExternalStorageProvider.
+ */
+fun isExternalStorageDocument(uri: Uri): Boolean {
+    return "com.android.externalstorage.documents" == uri.authority
+}
+
+/**
+ * @param uri The Uri to check.
+ * @return Whether the Uri authority is DownloadsProvider.
+ */
+fun isDownloadsDocument(uri: Uri): Boolean {
+    return "com.android.providers.downloads.documents" == uri.authority
+}
+
+/**
+ * @param uri The Uri to check.
+ * @return Whether the Uri authority is MediaProvider.
+ */
+fun isMediaDocument(uri: Uri): Boolean {
+    return "com.android.providers.media.documents" == uri.authority
+}
+
+/**
+ * @param uri The Uri to check.
+ * @return Whether the Uri authority is Google Photos.
+ */
+fun isGooglePhotosUri(uri: Uri): Boolean {
+    return "com.google.android.apps.photos.content" == uri.authority
+}
```

Comment:
I wonder where you need all these functions 🤔 
Also, this should go in data instead of presentation

Reply:
I will remove unneeded code. This was some code I was working with to get better handle on file uploading